File: frontend/src/api/notification.ts
const BASE = "/api/notifications";

export interface NotificationData {
  id: string;
  user_id: string;
  title: string;
  body: string;
  content?: string;
  level: "info" | "success" | "warning" | "error";
  agent_id?: string;
  agent_name?: string;
  read: boolean;
  created_at: string;
  images?: Array<{ data: string; media_type: string }>;
}

export async function fetchNotifications(
  limit = 50,
  unreadOnly = false
): Promise<NotificationData[]> {
  const params = new URLSearchParams({ limit: String(limit) });
  if (unreadOnly) params.set("unread_only", "true");
  const res = await fetch(`${BASE}?${params}`);
  if (!res.ok) throw new Error("Failed to fetch notifications");
  return res.json();
}

export async function fetchUnreadCount(): Promise<number> {
  const res = await fetch(`${BASE}/unread-count`);
  if (!res.ok) return 0;
  const data = await res.json();
  return data.count ?? 0;
}

export async function markNotificationRead(id: string): Promise<void> {
  await fetch(`${BASE}/${id}/read`, { method: "PUT" });
}

export async function markAllNotificationsRead(): Promise<void> {
  await fetch(`${BASE}/read-all`, { method: "PUT" });
}

export async function deleteNotification(id: string): Promise<void> {
  await fetch(`${BASE}/${id}`, { method: "DELETE" });
}

export async function clearAllNotifications(): Promise<void> {
  await fetch(BASE, { method: "DELETE" });
}

// ── Notification preferences ────────────────────────────────

export interface NotificationPreferences {
  delivery: "all" | "in_app" | "none";
}

export async function fetchNotificationPreferences(): Promise<NotificationPreferences> {
  const res = await fetch("/api/user/notification-preferences");
  if (!res.ok) return { delivery: "all" };
  return res.json();
}

export async function updateNotificationPreferences(
  prefs: NotificationPreferences
): Promise<NotificationPreferences> {
  const res = await fetch("/api/user/notification-preferences", {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(prefs),
  });
  if (!res.ok) throw new Error("Failed to update notification preferences");
  return res.json();
}

// ── Notification channels ──────────────────────────────────

export interface NotificationChannel {
  id: string;
  type: string;
  address: string;
  label: string;
  enabled: boolean;
}

export async function fetchNotificationChannels(): Promise<NotificationChannel[]> {
  const res = await fetch("/api/user/notification-channels");
  if (!res.ok) return [];
  return res.json();
}

export async function addNotificationChannel(
  channel: { type: string; address: string; label?: string }
): Promise<NotificationChannel> {
  const res = await fetch("/api/user/notification-channels", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(channel),
  });
  if (!res.ok) throw new Error("Failed to add channel");
  return res.json();
}

export async function updateNotificationChannel(
  id: string,
  updates: { label?: string; address?: string; enabled?: boolean }
): Promise<NotificationChannel> {
  const res = await fetch(`/api/user/notification-channels/${id}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(updates),
  });
  if (!res.ok) throw new Error("Failed to update channel");
  return res.json();
}

export async function deleteNotificationChannel(id: string): Promise<void> {
  await fetch(`/api/user/notification-channels/${id}`, { method: "DELETE" });
}

export async function testNotificationChannel(
  id: string
): Promise<{ success: boolean; message: string }> {
  const res = await fetch(`/api/user/notification-channels/${id}/test`, {
    method: "POST",
  });
  return res.json();
}

File: backend/app/routers/notification.py
"""
Notification API router.

Prefix: /api/notifications
"""

import logging
from fastapi import APIRouter, HTTPException

from app.services.notification_service import notification_service
from app.models.notification import NotificationResponse

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/notifications", tags=["notifications"])


@router.get("", response_model=list[NotificationResponse])
async def list_notifications(limit: int = 50, unread_only: bool = False):
    """List recent notifications, newest first."""
    items = notification_service.list_notifications(limit=limit, unread_only=unread_only)
    return [
        NotificationResponse(
            id=n["id"],
            user_id=n["user_id"],
            title=n["title"],
            body=n["body"],
            content=n.get("content", ""),
            level=n.get("level", "info"),
            agent_id=n.get("agent_id"),
            agent_name=n.get("agent_name"),
            read=n.get("read", False),
            created_at=n["created_at"],
            images=n.get("images"),
        )
        for n in items
    ]


@router.get("/unread-count")
async def unread_count():
    """Get the number of unread notifications."""
    count = notification_service.get_unread_count()
    return {"count": count}


@router.put("/{notification_id}/read")
async def mark_read(notification_id: str):
    """Mark a notification as read."""
    result = notification_service.mark_read(notification_id)
    if result is None:
        raise HTTPException(status_code=404, detail="Notification not found")
    return {"ok": True}


@router.put("/read-all")
async def mark_all_read():
    """Mark all notifications as read."""
    count = notification_service.mark_all_read()
    return {"ok": True, "count": count}


@router.delete("/{notification_id}")
async def delete_notification(notification_id: str):
    """Delete a notification."""
    ok = notification_service.delete_notification(notification_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Notification not found")
    return {"ok": True}


@router.delete("")
async def clear_all():
    """Delete all notifications."""
    count = notification_service.clear_all()
    return {"ok": True, "count": count}
